### PR TITLE
KIALI-1962 Suspend Traffic Wizard

### PR DESCRIPTION
**Pull request by lucasponce**
This PR introduces a new use-case Wizard for suspend traffic.
This operation might require different low level Istio configuration, depending if we want to stop traffic for a particular workload or whole service.

I.e. when more than one workload, typically a suspended traffic means to redirect traffic to pending workloads with weights. When all workloads are stopped, then a fault injection is generated.

The UI hides the complexity to the user and try to offer a one-click operation simplifying all steps around it.

https://issues.jboss.org/browse/KIALI-1962

image

image

image

image

### CODE REVIEW

**Comment by lucasponce** [comment]
> The 1 WL only case looks wrong. Shouldn't open not inject faults?

Oh, the screenshots are bad, as I posted them to show the style, but locally a 1 workload suspended should return a 503, and open do nothing.

> I'd also change "open" to "flowing" or "connected" to make it more clear what it means (think of a circuit breaker, where open means 'no connection').

+1 on "Connected"

**Comment by lucasponce** [comment]
@pilhuhn done
image

File: src/components/IstioWizards/IstioWizardDropdown.tsx [hunk 1]
```diff
@@ -172,6 +179,7 @@ class IstioWizardDropdown extends React.Component<Props, State> {
         <DropdownButton id="service_actions" title="Actions" onSelect={this.onAction} pullRight={true}>
           {this.canCreate() && this.renderMenuItem(WIZARD_WEIGHTED_ROUTING)}
           {this.canCreate() && this.renderMenuItem(WIZARD_MATCHING_ROUTING)}
+          {this.canCreate() && this.renderMenuItem(WIZARD_SUSPEND_TRAFFIC)}
```
Comment (xeviknal) [inline src/components/IstioWizards/IstioWizardDropdown.tsx]: nitpicking. What about running those three in a loop?
```typescript
WIZARDS.map((wizard) => this.canCreate() && this.renderMenuItem(wizard)
```

File: src/components/IstioWizards/IstioWizard.tsx [hunk 2]
```diff
@@ -216,6 +226,63 @@ class IstioWizard extends React.Component<Props, State> {
         };
         break;
       }
+      case WIZARD_SUSPEND_TRAFFIC: {
+        // VirtualService from the suspendedRoutes
+        const httpRoute: HTTPRoute = {
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply src/components/IstioWizards/IstioWizard.tsx]: I don't think it deserves such refactoring at the moment.
Main responsability of this component is to create the traffic using properties and state.
Split this function outside the component can be possible when it grows, but to me that adds complexity having the same logic I want to have in the same component.

File: src/components/IstioWizards/IstioWizard.tsx [hunk 3]
```diff
@@ -71,9 +78,12 @@ class IstioWizard extends React.Component<Props, State> {
           break;
         // By default no rules is a no valid scenario
         case WIZARD_MATCHING_ROUTING:
-        default:
           isValid = false;
           break;
+        case WIZARD_SUSPEND_TRAFFIC:
```
[the remark this answers is not in the document]
Comment (lucasponce) [reply src/components/IstioWizards/IstioWizard.tsx]: No. IstioWizard collects the state of the wizards, as it's responsability is to create the final configuration.

File: src/components/IstioWizards/IstioWizardDropdown.tsx (hunk 1 above)
Comment (lucasponce) [reply src/components/IstioWizards/IstioWizardDropdown.tsx]: On this case I prefer explictly name the options I guess, just to not hide it in a loop.

**Comment by hhovsepy** [comment]
[RFE] I think it would be useful from customers pov to be able to Suspend Traffic on existing Routing DR/VS, before going to delete them.
And the Wizard (probably in backend) will merge old VS traffic with the new one.
Because now user should delete the failing Routing (previously created), which can cause some more mess in traffic, then create a Suspend Routing.

**Comment by hhovsepy** [comment]
Suspending V1 traffic, the subsets for the V2 and V3 are gone, is it by design?
Screenshot from 2019-03-22 14-24-06

**Comment by lucasponce** [comment]
> Suspending V1 traffic, the subsets for the V2 and V3 are gone, is it by design?

Oh, no, that's a bug, thanks !

**Comment by lucasponce** [comment]
> [RFE] I think it would be useful from customers pov to be able to Suspend Traffic on existing Routing DR/VS, before going to delete them.
And the Wizard (probably in backend) will merge old VS traffic with the new one.
Because now user should delete the failing Routing (previously created), which can cause some more mess in traffic, then create a Suspend Routing.

Yes. That's included in the broader scope of work in the update of existing configuration from GUI perspective.

That's will be part of future https://issues.jboss.org/browse/KIALI-2573

**Comment by lucasponce** [comment]
@hhovsepy @xeviknal fixed last comment
image

**Pull request merged**